### PR TITLE
feat(dqlite): update workers to use new WatchMachines

### PR DESCRIPTION
**Pull request by cderici**
> [!NOTE]
> This was growing too much, so work is split out to https://github.com/juju/juju/pull/17558. This PR will be rebased onto that. The rest of description is kept for now to be used during the rebase and QA.

This adds the following:

- Updates `ModelMachinesWatcher` on the apiserver to use the new `WatchModelMachines` on the machine domain service.
- Updates the `provisioner`, `firewaller` and the `instancepoller` facades to pass the machine service to the ModelMachinesWorker.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

> [!CAUTION]
> Currently for a reason I've yet to understand, this causes firewaller to bork after bootstrap with:
``` $ jdebug -m controller
unit-controller-0: 14:52:33 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
controller-0: 14:52:41 INFO juju.worker.firewaller runner is dying
machine-0: 14:52:41 ERROR juju.worker.dependency "firewaller" manifold worker returned unexpected error: permission denied 
```

## QA steps

All following should work on lxd and aws.

```
 $ juju bootstrap lxd
 $ juju add-model deleteme
 $ juju deploy ubuntu
 $ juju add-machine -n 2
 $ juju add-machine lxd:1
```

Also some unit tests of interest:

```
 $ TEST_PACKAGES="./domain/schema/... -count=1 -race" make run-go-tests
 $ TEST_PACKAGES="./domain/machine/... -count=1 -race" make run-go-tests
 $ TEST_PACKAGES="./apiserver/facades/controller/firewaller/... -count=1 -race" make run-go-tests
 $ TEST_PACKAGES="./apiserver/facades/controller/instancepoller/... -count=1 -race" make run-go-tests
```

## Links

**Jira card:** JUJU-

### CODE REVIEW

**Comment by hpidcock** [comment]
The provisioner worker is run on machines, they cannot access the service factory. We need to continue to use the facade calls here.

**Review by SimonRichardson — CHANGES_REQUESTED** [review]
@hpidcock is indeed correct here. You can't use the services directly in the provisioner. The error you got is on purpose. Essentially, you're trying to import the database into `jujud`, which has no concept of a database and can only communicate to the database via the API server, which lives in `jujud-controller`.

There are two solutions:

 1. Go via the API server (facades) for all calls.
 2. For jujud-controller provisioner, use the service factory directly and for jujud go via the API server.

File: domain/machine/watcher_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+func (s *watcherSuite) TestWatchWithDelete(c *gc.C) {
+	machineService, watcherC := s.machineServiceAndWatcher(c)
+
+	// Create a machine
+	uuid, err := machineService.CreateMachine(context.Background(), "machine-1")
+	c.Assert(err, gc.IsNil)
+	// Delete the machine
+	err = machineService.DeleteMachine(context.Background(), "machine-1")
+	c.Assert(err, gc.IsNil)
+	watcherC.AssertChange(uuid)
```
Comment (SimonRichardson) [inline domain/machine/watcher_test.go:45]: This isn't quite correct, instead I'd expect one change when we create machine, then one for when we delete the machine.

```suggestion
	// Create a machine
	uuid, err := machineService.CreateMachine(context.Background(), "machine-1")
	c.Assert(err, gc.IsNil)
	
	// Ensure that the changestream is idle.
	s.ModelSuite.AssertChangeStreamIdle(c)
	
	// Assert the first change
	watcherC.AssertChange(uuid)
	
	// Delete the machine
	err = machineService.DeleteMachine(context.Background(), "machine-1")
	c.Assert(err, gc.IsNil)

	// Ensure that the changestream is idle.
	s.ModelSuite.AssertChangeStreamIdle(c)
	
	watcherC.AssertChange(uuid)
```

File: domain/machine/state/state.go [hunk 2]
```diff
@@ -81,6 +82,7 @@ VALUES ($M.machine_uuid, $M.net_node_uuid, $M.machine_id, $M.life_id)
 		if err != nil {
 			return errors.Trace(err)
 		}
+		newUUID = machineUUID.String()
```
Comment (manadart) [inline domain/machine/state/state.go:85]: We should be doing this up in the service layer. If retries are invoked, we're generating them over again.

**Review by manadart — COMMENTED** [review]
You're trying to do too much at once here.

Land the domain changes first, along with the signature changes that include the UUID.

I'm pretty sure the error is caused by your watcher returning machine UUIDs, where the old one returned _machine_ IDs. When you get the change and try to query them, they are not found.

The firewaller and instance-poller should be accessing the service directly as we discussed before. It's only the on-machine provisioner that needs to use the API.

**Comment by cderici** [comment]
Closing this in favor of a new one. 1) it's redundant to try to figure out the conflicts (things have moved around), 2) firewaller and instancepoller needs to access the services directly so those parts will be newly written anyways.

**Pull request closed without merging**